### PR TITLE
watchtower api

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---
Policing service grpc api
[Client / Server interaction diagram](https://docs.google.com/presentation/d/1MkLraDsZJCX3k62bBtt8ziUdw308U2bW_w0lUmV-0YQ/edit?usp=sharing)

### CODE REVIEW

**Comment by shayzluf** [comment]
> Why return slashing containers rather than bool. What advantages do we have for returning the containers here?

This is a way to inject data as well as to validate that the data that is being sent is not creating a slashable event. i thought the injector should be the first to get the slashable event as a response

File: scripts/init_keystore.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
```
Comment (nisdas) [inline scripts/init_keystore.sh]: why is this needed ?

File: proto/beacon/p2p/v1/messages.pb.go [hunk 2]
```diff
@@ -5,12 +5,11 @@ package ethereum_beacon_p2p_v1
...
 import (
 	fmt "fmt"
-	io "io"
-	math "math"
-
 	proto "github.com/gogo/protobuf/proto"
 	types "github.com/gogo/protobuf/types"
 	v1alpha1 "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	io "io"
+	math "math"
```
Comment (terencechain) [inline proto/beacon/p2p/v1/messages.pb.go]: why does this happen, do you wanna revert this?

File: scripts/init_keystore.sh (hunk 1 above)
Comment (terencechain) [reply scripts/init_keystore.sh]: +1

File: proto/sharding/watchtower/watchtower.proto [hunk 3]
```diff
@@ -0,0 +1,32 @@
+syntax = "proto3";
```
Comment (rauljordan) [inline proto/sharding/watchtower/watchtower.proto]: Can we move this file to the proto/eth folder?

File: proto/sharding/watchtower/watchtower.proto [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+// attestation or proposal they are going to submit to the network are not going to 
+// produce a slashable event.
+service Watchtower {
+    
+    // Gets AttesterSlashing container if the attestation that 
+    // was recieved produces a slashable event.
+    rpc IsSlashableAttestation(ethereum.eth.v1alpha1.Attestation) returns (ethereum.eth.v1alpha1.AttesterSlashing) ;
+
+    // Gets ProposerSlashing container if the block header that 
+    // was recieved produces a slashable event.
```
Comment (rauljordan) [inline proto/sharding/watchtower/watchtower.proto]:
```suggestion
    // was received produces a slashable event.
```

File: proto/sharding/watchtower/watchtower.proto [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// Watchtower service API
+// 
+// Watchtower service provides an interface for validators and beacon server to query
+// to subscribe for slashable events on the network as well as to make sure that the 
+// attestation or proposal they are going to submit to the network are not going to 
+// produce a slashable event.
+service Watchtower {
+    
+    // Gets AttesterSlashing container if the attestation that 
+    // was recieved produces a slashable event.
```
Comment (rauljordan) [inline proto/sharding/watchtower/watchtower.proto]:
```suggestion
    // was received produces a slashable event.
```

File: proto/sharding/watchtower/watchtower.proto [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    rpc IsSlashableAttestation(ethereum.eth.v1alpha1.Attestation) returns (ethereum.eth.v1alpha1.AttesterSlashing) ;
+
+    // Gets ProposerSlashing container if the block header that 
+    // was recieved produces a slashable event.
+    rpc IsSlashableBlock(ethereum.eth.v1alpha1.BeaconBlockHeader) returns (ethereum.eth.v1alpha1.ProposerSlashing) ;
+
+    // Subscription to recieve all slashable proposer slashing events found by the watchtower.
```
Comment (rauljordan) [inline proto/sharding/watchtower/watchtower.proto]:
```suggestion
    // Subscription to receive all slashable proposer slashing events found by the watchtower.
```

File: proto/sharding/watchtower/watchtower.proto [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    rpc SlashableProposals(google.protobuf.Empty) returns (stream ethereum.eth.v1alpha1.ProposerSlashing);
+    
+    // Subscription to recieve all slashable attester slashing events found by the watchtower.
```
Comment (rauljordan) [inline proto/sharding/watchtower/watchtower.proto]:
```suggestion
    // Subscription to receive all slashable attester slashing events found by the watchtower.
```

File: proto/beacon/p2p/v1/types.pb.go [hunk 8]
```diff
@@ -5,13 +5,12 @@ package ethereum_beacon_p2p_v1
...
 	fmt "fmt"
-	io "io"
-	math "math"
-
 	_ "github.com/gogo/protobuf/gogoproto"
 	proto "github.com/gogo/protobuf/proto"
 	github_com_prysmaticlabs_go_bitfield "github.com/prysmaticlabs/go-bitfield"
 	v1alpha1 "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	io "io"
+	math "math"
```
Comment (0xKiwi) [inline proto/beacon/p2p/v1/types.pb.go]: These imports should be with the top ones like the other native imports.

File: scripts/init_keystore.sh (hunk 1 above)
Comment (shayzluf) [reply scripts/init_keystore.sh]: removed, thanks

File: proto/sharding/watchtower/watchtower.proto (hunk 3 above)
Comment (shayzluf) [reply proto/sharding/watchtower/watchtower.proto]: inside v1alpha? or inside watchtower dir?

File: proto/sharding/watchtower/watchtower.proto (hunk 4 above)
Comment (shayzluf) [reply proto/sharding/watchtower/watchtower.proto]: fixed

File: proto/sharding/watchtower/watchtower.proto (hunk 5 above)
Comment (shayzluf) [reply proto/sharding/watchtower/watchtower.proto]: fixed

File: proto/beacon/p2p/v1/messages.pb.go (hunk 2 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/messages.pb.go]: pbs generation did not got fix by goimport

Comment (shayzluf) [reply proto/beacon/p2p/v1/messages.pb.go]: reverted

File: proto/sharding/watchtower/watchtower.proto (hunk 3 above)
Comment (shayzluf) [reply proto/sharding/watchtower/watchtower.proto]: moved

File: proto/sharding/watchtower/watchtower.proto (hunk 6 above)
Comment (shayzluf) [reply proto/sharding/watchtower/watchtower.proto]: thanks, fixed

File: proto/sharding/watchtower/watchtower.proto (hunk 7 above)
Comment (shayzluf) [reply proto/sharding/watchtower/watchtower.proto]: fixed

File: proto/beacon/p2p/v1/types.pb.go (hunk 8 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/types.pb.go]: fixed

File: proto/eth/v1alpha1/slasher.proto [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+// Slasher service API
+// 
+// Slasher service provides an interface for validators and beacon server to query
+// to subscribe for slashable events on the network as well as to make sure that the 
+// attestation or proposal they are going to submit to the network are not going to 
+// produce a slashable event.
+service Slasher {
+
+
```
Comment (rauljordan) [inline proto/eth/v1alpha1/slasher.proto]: Eliminate the white space here and line 18

File: proto/eth/v1alpha1/slasher.proto [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+service Slasher {
+
+    
+    // Gets AttesterSlashing container if the attestation that 
+    // was received produces a slashable event.
+    rpc IsSlashableAttestation(ethereum.eth.v1alpha1.Attestation) returns (ethereum.eth.v1alpha1.AttesterSlashing) ;
+
+    // Gets ProposerSlashing container if the block header that 
+    // was received produces a slashable event.
+    rpc IsSlashableBlock(ethereum.eth.v1alpha1.BeaconBlockHeader) returns (ethereum.eth.v1alpha1.ProposerSlashing) ;
```
Comment (rauljordan) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
    rpc IsSlashableBlock(ethereum.eth.v1alpha1.BeaconBlockHeader) returns (ethereum.eth.v1alpha1.ProposerSlashing);
```

File: proto/eth/v1alpha1/slasher.proto [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+// Slasher service provides an interface for validators and beacon server to query
+// to subscribe for slashable events on the network as well as to make sure that the 
+// attestation or proposal they are going to submit to the network are not going to 
+// produce a slashable event.
+service Slasher {
+
+    
+    // Gets AttesterSlashing container if the attestation that 
+    // was received produces a slashable event.
+    rpc IsSlashableAttestation(ethereum.eth.v1alpha1.Attestation) returns (ethereum.eth.v1alpha1.AttesterSlashing) ;
```
Comment (rauljordan) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
    rpc IsSlashableAttestation(ethereum.eth.v1alpha1.Attestation) returns (ethereum.eth.v1alpha1.AttesterSlashing);
```

File: proto/eth/v1alpha1/slasher.proto (hunk 9 above)
Comment (rauljordan) [reply proto/eth/v1alpha1/slasher.proto]: No need to regen protos after that

File: proto/eth/v1alpha1/slasher.proto (hunk 11 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: tx

File: proto/eth/v1alpha1/slasher.proto (hunk 10 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: tx

File: proto/eth/v1alpha1/slasher.proto (hunk 9 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: removed

File: proto/eth/v1alpha1/slasher.proto [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+import "google/protobuf/empty.proto";
+import "proto/eth/v1alpha1/beacon_block.proto";
+import "proto/eth/v1alpha1/attestation.proto";
+
+option go_package = "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1;eth";
+
+// Slasher service API
+// 
+// Slasher service provides an interface for validators and beacon server to query
```
Comment (terencechain) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
// Slasher service provides an interface for validators and beacon chain server to query
```

File: proto/eth/v1alpha1/slasher.proto [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+// to subscribe for slashable events on the network as well as to make sure that the
```
Comment (terencechain) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
// and subscribe for slashable events on the network as well as to make sure that the 
```
